File: Patterns/Strategy Pattern/Cpp/Duck/Duck.cpp
#include <iostream>
#include "../Behavior/IFlyBehavior.cpp"
#include "../Behavior/IQuackBehavior.cpp"

using namespace std;

class Duck {
protected:
	IFlyBehavior* flyBehavior;
	IQuackBehavior* quackBehavior;

public:
	Duck() { }

	virtual void display() = 0;

	void swim();

	void performFly();
	void performQuack();

	void setFlyBehavior(IFlyBehavior* fb);
	void setQuackBehavior(IQuackBehavior* qb);
};

void Duck::swim() {
	cout << "I Can Swim.\n";
}

void Duck::performFly() {
	flyBehavior->fly();
}

void Duck::performQuack() {
	quackBehavior->quack();
}

void Duck::setFlyBehavior(IFlyBehavior* fb) {
	flyBehavior = fb;
}

void Duck::setQuackBehavior(IQuackBehavior* qb) {
	quackBehavior = qb;
}

File: Patterns/Strategy Pattern/CSharp/Duck/Duck.cs
using System;

using IFlyBehavior = Behavior.IFlyBehavior;
using IQuackBehavior = Behavior.IQuackBehavior;

namespace Duck 
{
    public abstract class Duck 
    {
        protected IFlyBehavior flyBehavior; 
        protected IQuackBehavior quackBehavior;

        public Duck(IFlyBehavior fb, IQuackBehavior qb) 
        {
            flyBehavior = fb;
            quackBehavior = qb;
        }

        public abstract void OnDisplay();

        public void Swim() => Console.WriteLine("I Can Swim.");
        
        public void PerformFly() => flyBehavior.OnFly();
        public void PerformQuack() => quackBehavior.OnQuack();
    }
}